General method for modifying printing of elements in arrays

gajomi opened the pull request:
Added logic to Base.print_matrix_row and introduced isstructuralzero methods to for sparse pretty printing of matrices with structural zeros.

hayd commented:
This could do with some tests!

tkelman commented on test/show.jl:
```diff
@@ -313,3 +313,13 @@ end
+
+# test structured zero matrix printing for select structured types
+A = reshape(1:16,4,4)
+@test replstr(Diagonal(A)) == "4x4 Diagonal{Int64}:\n 1           \n    6        \n       11    \n           16"
```
should use `$Int` instead of `Int64` so this works on 32 bit

gajomi replied:
Thanks. Made the change.

jakebolewski commented:
I think something should be printed in the structural zero location, maybe cdot (`⋅`)? Otherwise it is hard to infer the structure for larger matrices.

StefanKarpinski commented:
How about printing `zero(T)` in the structural zero locations?

jakebolewski commented:
I don't understand, that is what is printed now?

KristofferC commented:
I guess that was a joke :p

StefanKarpinski commented:
"Structural zeros" is kind of an ambiguous term – is it the stored or unstored zeros? I'm proposing that stored zeros print as `0.0` and unstored zeros print as nothing. Are you talking about stored zeros or unstored zeros?

tkelman commented:
Structural zero means the implicit non-stored zeros in my mind, but if it's confusable then we should use a different name. Go the affirmative direction and use `isstored`?

jakebolewski commented:
Agree that the name is not good, just going along with terminology used here.

While this is fine for small examples, I think this concept falls down when the matrix spans a width greater than the terminal.

```julia
julia> Diagonal(rand(10,10))
10x10 Diagonal{Float64}:
 0.729744
           0.906219
                     0.118884
                               0.299574
                                         0.361695
                                                   0.921484
                                                             0.692415
                                                                       0.288065
                                                                                 0.312887
                                                                                           0.0647774

julia> Diagonal(rand(100,100))
100x100 Diagonal{Float64}:
 0.183967                                                     …
           0.0478895
                      0.495338
                                0.924295
                                          0.910174
                                                    0.153329  …




                                                              …




                                                              …




                                                              …


 ⋮                                                  ⋮         ⋱  ⋮


                                                              …




                                                              …




                                                              …




                                                              …  0.604073
                                                                           0.638792
                                                                                     0.272355
                                                                                               0.739478
                                                                                                         0.296522
```

StefanKarpinski commented:
When it spans a width wider than the terminal, it seems like it should elide the matrix like we currently do.

gajomi commented:
@StefanKarpinski , `isstructuralzero` yeah not the best name. The idea is as you say: the elements in question are not stored and are equal to zero. So, following @tkelman something like `isnonstoredzero` might be better.

gajomi commented:
@jakebolewski +1 on cdots. Here is what it looks like with that change:

```julia
4x4 Diagonal{Float64}:
 0.125748   ⋅         ⋅         ⋅      
  ⋅        0.867896   ⋅         ⋅      
  ⋅         ⋅        0.795273   ⋅      
  ⋅         ⋅         ⋅        0.719751
 
4x4 Tridiagonal{Float64}:
 0.943428  0.588128    ⋅         ⋅      
 0.39349   0.0542561  0.469953   ⋅      
  ⋅        0.121129   0.707122  0.530904
  ⋅         ⋅         0.19026   0.842388
 
4x4 Bidiagonal{Float64}:
 0.0250798  0.983997   ⋅         ⋅      
  ⋅         0.879443  0.693986   ⋅      
  ⋅          ⋅        0.429402  0.449712
  ⋅          ⋅         ⋅        0.169836
 
4x4 UpperTriangular{Float64,Array{Float64,2}}:
 0.773531  0.641691  0.372334  0.144952
  ⋅        0.609429  0.652985  0.691839
  ⋅         ⋅        0.679731  0.972232
  ⋅         ⋅         ⋅        0.295862
 
4x4 LowerTriangular{Float64,Array{Float64,2}}:
 0.222366   ⋅         ⋅         ⋅      
 0.15382   0.919547   ⋅         ⋅      
 0.881704  0.307257  0.328537   ⋅      
 0.752798  0.393471  0.378299  0.628847
```

And for your examples
```julia
julia> Diagonal(rand(10,10))
10x10 Diagonal{Float64}:
 0.573221   ⋅         ⋅         ⋅         ⋅         ⋅         ⋅         ⋅         ⋅         ⋅      
  ⋅        0.480602   ⋅         ⋅         ⋅         ⋅         ⋅         ⋅         ⋅         ⋅      
  ⋅         ⋅        0.913413   ⋅         ⋅         ⋅         ⋅         ⋅         ⋅         ⋅      
  ⋅         ⋅         ⋅        0.680573   ⋅         ⋅         ⋅         ⋅         ⋅         ⋅      
  ⋅         ⋅         ⋅         ⋅        0.138932   ⋅         ⋅         ⋅         ⋅         ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅        0.778682   ⋅         ⋅         ⋅         ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅        0.352086   ⋅         ⋅         ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅         ⋅        0.706759   ⋅         ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅         ⋅         ⋅        0.626003   ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅         ⋅         ⋅         ⋅        0.306449

julia> Diagonal(rand(100,100))
100x100 Diagonal{Float64}:
 0.362967   ⋅         ⋅         ⋅         ⋅         ⋅        …   ⋅         ⋅         ⋅         ⋅         ⋅        ⋅      
  ⋅        0.483187   ⋅         ⋅         ⋅         ⋅            ⋅         ⋅         ⋅         ⋅         ⋅        ⋅      
  ⋅         ⋅        0.194026   ⋅         ⋅         ⋅            ⋅         ⋅         ⋅         ⋅         ⋅        ⋅      
  ⋅         ⋅         ⋅        0.127909   ⋅         ⋅            ⋅         ⋅         ⋅         ⋅         ⋅        ⋅      
  ⋅         ⋅         ⋅         ⋅        0.278139   ⋅            ⋅         ⋅         ⋅         ⋅         ⋅        ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅        0.492785  …   ⋅         ⋅         ⋅         ⋅         ⋅        ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅            ⋅         ⋅         ⋅         ⋅         ⋅        ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅            ⋅         ⋅         ⋅         ⋅         ⋅        ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅            ⋅         ⋅         ⋅         ⋅         ⋅        ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅            ⋅         ⋅         ⋅         ⋅         ⋅        ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅        …   ⋅         ⋅         ⋅         ⋅         ⋅        ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅            ⋅         ⋅         ⋅         ⋅         ⋅        ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅            ⋅         ⋅         ⋅         ⋅         ⋅        ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅            ⋅         ⋅         ⋅         ⋅         ⋅        ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅            ⋅         ⋅         ⋅         ⋅         ⋅        ⋅      
 ⋮                                                 ⋮         ⋱            ⋮                                              
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅        …   ⋅         ⋅         ⋅         ⋅         ⋅        ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅            ⋅         ⋅         ⋅         ⋅         ⋅        ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅            ⋅         ⋅         ⋅         ⋅         ⋅        ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅            ⋅         ⋅         ⋅         ⋅         ⋅        ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅            ⋅         ⋅         ⋅         ⋅         ⋅        ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅        …   ⋅         ⋅         ⋅         ⋅         ⋅        ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅            ⋅         ⋅         ⋅         ⋅         ⋅        ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅            ⋅         ⋅         ⋅         ⋅         ⋅        ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅            ⋅         ⋅         ⋅         ⋅         ⋅        ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅           0.761387   ⋅         ⋅         ⋅         ⋅        ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅        …   ⋅        0.187665   ⋅         ⋅         ⋅        ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅            ⋅         ⋅        0.432716   ⋅         ⋅        ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅            ⋅         ⋅         ⋅        0.569741   ⋅        ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅            ⋅         ⋅         ⋅         ⋅        0.23445   ⋅      
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅            ⋅         ⋅         ⋅         ⋅         ⋅       0.717876
```

Its nice in that its still lightweight but makes it easier to count out where you are in the structure. The oversized matrix starts to look a bit heavier, but is significantly lighter than the currect behavior explicitly printing non-stored zeros.

tkelman commented:
asking in the affirmative is generally more direct than asking `isnotwhatever`

jakebolewski commented:
i think that looks pretty good.

gajomi commented:
@tkelman - Yeah I felt awkward when writing that. `isstored` is probably better. I hesitate to use that exact name though, since there are special matrix types with non-stored values that are also nonzero. But I might be overthinking it.

gajomi commented:
@StefanKarpinski - The current eliding should be the same as before. 

Compare:
```julia
julia> Diagonal(ones(100))
100x100 Diagonal{Float64}:
 1.0   ⋅    ⋅    ⋅    ⋅    ⋅   …   ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅   1.0   ⋅    ⋅    ⋅    ⋅       ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅   1.0   ⋅    ⋅    ⋅       ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅   1.0   ⋅    ⋅       ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅   1.0   ⋅       ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅   1.0  …   ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅       ⋅    ⋅    ⋅    ⋅    ⋅ 
 ⋮                        ⋮    ⋱  ⋮                      
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅       ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅       ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   …  1.0   ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅       ⋅   1.0   ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅       ⋅    ⋅   1.0   ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅       ⋅    ⋅    ⋅   1.0   ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅       ⋅    ⋅    ⋅    ⋅   1.0
```
to
```julia
julia> full(Diagonal(ones(100)))
100x100 Array{Float64,2}:
 1.0  0.0  0.0  0.0  0.0  0.0  …  0.0  0.0  0.0  0.0  0.0
 0.0  1.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0
 0.0  0.0  1.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  1.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  1.0  0.0     0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  1.0  …  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0
 ⋮                        ⋮    ⋱  ⋮                      
 0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0  …  1.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0     0.0  1.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  1.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  1.0  0.0
 0.0  0.0  0.0  0.0  0.0  0.0     0.0  0.0  0.0  0.0  1.0
```

gajomi commented:
If people are happy with cdots i'll cook up some new tests and commit the changes.

tkelman commented:
should first check if cdot displays properly in the windows console. might show up as question marks or boxes which would confuse windows users.

nalimilan commented:
Looks nice, but IMHO it would be even better without the lines containing only dots, so that the oblique series of dots align with the diagonal, i.e.:
```
 1.0   ⋅    ⋅    ⋅    ⋅    ⋅   …   ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅   1.0   ⋅    ⋅    ⋅    ⋅       ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅   1.0   ⋅    ⋅    ⋅       ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅   1.0   ⋅    ⋅       ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅   1.0   ⋅       ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅   1.0  …   ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋮                       ⋮  ⋱   ⋮                      
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   …  1.0   ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅       ⋅   1.0   ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅       ⋅    ⋅   1.0   ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅       ⋅    ⋅    ⋅   1.0   ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅       ⋅    ⋅    ⋅    ⋅   1.0
```

KristofferC commented:
This is how it looks in my terminal. How picky are we in general with imposing a fully unicode compliant font on users?

![image](https://cloud.githubusercontent.com/assets/1282691/10758255/97d0c040-7cb0-11e5-82d9-5a7a453993b0.png)

jiahao commented:
Very ;-)

More seriously, it looks like your terminal font is missing a glyph for the vertical dots character ⋮ , but we use it for truncation already (you should see it used in displaying `randn(100,100)`), so I don't think this PR can be blamed for using this character.

KristofferC commented:
Ah, hadn't noticed that. Carry on then :)

mmoh commented:
On a related note I can't get the vertical dots and diagonal dots using DejaVuSans Mono. (using conemu, but same observations using consolez)

jiahao commented:
@mmoh DejaVu Sans Mono does not currently have those glyphs, so you cannot use this terminal font if you want the dots to render.

mmoh commented:
@jiahao what font do you recommend to get for all the unicode glyphs support?

jakebolewski commented:
gnu unifont

On Tue, Oct 27, 2015 at 10:30 AM, mmoh <notifications@github.com> wrote:

> @jiahao <https://github.com/jiahao> what font do you recommend to get for
> all the unicode glyphs support?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaLang/julia/pull/13732#issuecomment-151517109>.
>

gajomi commented:
cdots in last commit. I decided to go with `isunstoredzero` as the new method name. In my particular sense for English this was far less jarring than `isnonstoredzero` but still explicit about the exact concept behind the name.

tkelman commented:
We're not exporting it just yet here, but thinking about the general concept that this verb should represent, it could extend beyond just types where the implicit non stored values are equal to zero. In combination with a future `defaultvalue` or similar method I think `isstored` would be the more generic concept here.

StefanKarpinski commented:
This does bear some more careful consideration. Which values, if any, of a range object are stored?

gajomi commented:
@tkelman - Yeah I agree this is worth thinking about for possible uses beyond this printing issue.  From the SpecialMatrices.jl packages we have Hilbert and Riemann Matrices and, which are nonzero at every entry but only store their size. One could similarly imagine a simple sized parameterized immutable type for a DFT matrix. As @StefanKarpinski points out, you also have things like ranges, which need to store information beyond just their size, but are still will need to calculate the values particular indexes on demand. To further muddy the waters you have things with permutation symmetries like Toeplitz matrices in which the returned values don't need to be calculated and are stored, but in a container of size no greater than O(N), which has different requirements than the dense case. So, in terms of categories of matrix elements, I can see some reason to distinguish between zero/nonzero, stored/unstored and calculated/retrieved cases.

The printing of unstored sparsity issue here is seemingly about the first two categories, and perhaps suggests that composition of two methods (say `iszero` and `isstored`) might make for a cleaner way api.

StefanKarpinski commented:
You could also represent sparse data structures using something like run-length encoded vectors instead of having a fixed default. In many ways this is much more flexible than sparse structures with a fixed default and not that much less efficient. What qualifies as a stored value then?

gajomi commented:
I'm not sure I know how to draw the line around everything I would say to be a "stored value" although I would consider values retrieved from a run-length encoding to be examples of them. 

But it just occurred to me that perhaps, wrt to the print issue, it might be better to set aside some of the semantics of the method(s) we are talking about. Namely, we instead define an 'arrayprintcharacter(A,i,j)` method or something along those lines, that returned something falsey for the default behavior and something truthy like cdots (or whatever appropriate length string that user wanted) to override the default. In this way, the cdots need not have any enforced meaning, but we would advocate for the convention of using them to denote unstored zeros, whatever we mean by that :).

tkelman commented:
Good idea, focusing on a printing-specific meaning and name for now is probably a good idea to move forward here without getting too far into the traits weeds.

toivoh commented:
But if we use cdot to signify some zeros, I really think we should settle
on the convention that it will only signify zeros. So for the printing
purposes at hand, maybe it really is sufficient to be able to distinguish
structural (non-stored) zeros from anything else.

gajomi commented:
@toivoh - I fully agree, cdots should always be for zeros (unless its a matrix of utf8 chars, then they are just being). 

It is interesting to think though about things beyond cdots for other purposes. You essentially have a fairly general hook into the printing of elements for matrices/vector. This could definitely be abused. But, to give an example of possibly interesting use...

Imagine you are designing a sieve and want to visualize it in action. So you define a FunPrimeMat type, with a method that blackboard the prime entries to et:
```julia
6x6 FunPrimeMat:
 𝟟   8   9  10  𝟙𝟙  12
𝟙𝟛  14  15  16  𝟙𝟟  18
𝟙𝟡  20  21  22  𝟚𝟛  24
25  26  27  28  𝟚𝟡  30
𝟛𝟙  32  33  34  35  36
𝟛𝟟  38  39  40  41  42
```
which is nice for finding certain patterns. Or for a more complicated version you can imagine the same on a [Ulam spiral](http://mathworld.wolfram.com/PrimeSpiral.html) type. 

This is pretty academic, but I imagine Sudoku aficionados could cook up some more relevant use cases.

StefanKarpinski commented:
The sensible approach here seems to me to be to use the `⋅` for entries in numeric matrices which are zero-valued *and* "structurally uninteresting" – which is an intentionally vague concept that depends on the matrix type. For dense matrices, no entries would ever satisfy this since there are no structurally uninteresting locations. For diagonal matrices, the off-diagonal could be printed as `⋅` but zeros on the diagonal would not be.

gajomi commented:
I was originally a bit hesitant to go with a general hook into the printing of matrix elements, as described above, since it definitely could be abused. But I eventually realized that there was no more opportunity for abuse then already existed with the possibility of defining new `display` methods. And so I will describe the behavior in the last commit. The (unexported) basic hook method is an `replace_in_print_matrix(A::FunPrimeMat,i::Integer,j::Integer,s::AbstractString)`
which will dispatch on custom array types and can use any information about the array, the index and the default printing string `s` to return a new string to be printed. The canonical case is to add cdots wherever there are "structural uninteresting zeros" (I endorse @StefanKarpinski's verbage). The behavior for the types in Julia is as before. To get this behavior in a custom type one can make use of the (unexported) `replace_with_centered_cdots(s::AbstractString)` method as follows
```julia
using SpecialMatrices

function Base.replace_in_print_matrix(X::Frobenius,i::Integer,j::Integer,s::AbstractString)
  i==j || j==X.colidx && j<i ? s : Base.replace_with_centered_cdots(s)
end
```
to get
```julia
julia> F=Frobenius(3, collect(1.0:3.0))
6x6 SpecialMatrices.Frobenius{Float64}:
 1.0   ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅   1.0   ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅   1.0   ⋅    ⋅    ⋅ 
  ⋅    ⋅   1.0  1.0   ⋅    ⋅ 
  ⋅    ⋅   2.0   ⋅   1.0   ⋅ 
  ⋅    ⋅   3.0   ⋅    ⋅   1.0
```
This is (in the scope of this PR) the standard use case. Their is a bit of DRY violation in the implementations, but it is a least explicit about what is happening. 

The hook allows for more exotic things though. For example, as was eluded to before one can write things like:
```julia
type FunPrimeMat <: AbstractMatrix{Int64}
  N::Int64
end

Base.size(A::FunPrimeMat) = (A.N,A.N)
Base.getindex(A::FunPrimeMat,i,j) = A.N*(i-1)+j

istwinprime(n::Integer) = isprime(n) && (isprime(n+2) || isprime(n-2))

function Base.replace_in_print_matrix(A::FunPrimeMat,i::Integer,j::Integer,s::AbstractString)
  bbfun =   c->Dict(zip(split("0123456789",""),split("𝟘𝟙𝟚𝟛𝟜𝟝𝟞𝟟𝟠𝟡","")))[string(c)]
  boldfun = c->Dict(zip(split("0123456789",""),split("𝟎𝟏𝟐𝟑𝟒𝟓𝟔𝟕𝟖𝟗","")))[string(c)]
  x = A[i,j]
  if isprime(x)
    if istwinprime(x)
      return replace(s,r"[0-9]",boldfun)
    else
      return replace(s,r"[0-9]",bbfun)
    end
  else
    return s
  end
end
```
which gives you printing like
```julia
julia> A = FunPrimeMat(6)
6x6 FunPrimeMat:
  1   𝟚   𝟑   4   𝟓   6
  𝟕   8   9  10  𝟏𝟏  12
 𝟏𝟑  14  15  16  𝟏𝟕  18
 𝟏𝟗  20  21  22  𝟚𝟛  24
 25  26  27  28  𝟐𝟗  30
 𝟑𝟏  32  33  34  35  36
```

To be clear, I am not specifically advocating for using my stupid `FunPrimeMat` type or highlighting primes and the like. I just wish to illustrate that the `replace_in_print_matrix` leaves the possibility open. This is useful for users who want to get custom display behavior on elements, but want to still have the default header, footer and alignment behavior of Matrix/Vector printing.

gajomi commented:
I have modified the PR title to reflect the expanded scope.

artkuo commented:
I like the cdots. I suggest rather than hard-coding cdots, make that an optional parameter, as in `cdot::AbstractString = "⋅"`. This would alllow someone to use space if they prefer, and it would be consistent with the related methods, e.g. `print_matrix` takes in optionals for vdots, hdots, separators, etc. If applied here, use `length(cdot)` in case the string is more than one character, but otherwise it seems easy. Also, `print_matrix` is getting long in parameters, so consider making the strings unordered optionals with semicolon.

gajomi commented:
@artkuo Good suggestion. I made the change in last commit so you can now put 'x' for '☠' or whatever pleases you. If anyone needs more general behavior they can implement custom `replace_in_print_matrix` methods.

andreasnoack commented:
I like this. Can we merge (after a rebase)?

tkelman commented on base/linalg/bidiag.jl:
```diff
@@ -40,6 +40,15 @@ function getindex{T}(A::Bidiagonal{T}, i::Integer, j::Integer)
 
+## structured matrix methods ##
+function Base.replace_in_print_matrix(A::Bidiagonal,i::Integer,j::Integer,s::AbstractString)
+  if A.isupper
```
also fix the indent to be 4-spaced

gajomi replied:
Guess I hadn't read about that particular convention just yet. I'll make the changes by hand for now, but it would be nice to have something automatic. Is there an official linter that conforms to the style in the manual so I can write a plugin for my editor?

tkelman replied:
ref https://github.com/JuliaLang/julia/blob/master/CONTRIBUTING.md#code-formatting-guidelines

What editor are you using? 4-space indent is a good default to set for contributing to Julia, and most packages follow the same convention.

gajomi replied:
I just made the changes now in nano (I'm in the boondocks on another machine), but I wrote everything before in atom.

tkelman replied:
Heh. At the risk of having my commit access taken away, I totally use nano too when I'm in a terminal.

Formatting looks good now, thanks, but it does still need a rebase to fix a merge conflict.

gajomi commented:
Hmm, well rebase ran successfully on my machine without conflicts. Am I trying to rebase into the wrong master? I am a git novice, so I might need a pointer about how to reproduce the merge problem. 

Here is what I did:
```
> git clone https://github.com/gajomi/julia
> git checkout structuredmatrixprinting
...
> #made the changes to files and committed
...
> git rebase -i master #left everything as it was and ran rebase
Successfully rebased and updated refs/heads/structuredmatrixprinting.
```

tkelman commented:
You should add JuliaLang as a remote:
```sh
git remote add JuliaLang https://github.com/JuliaLang/julia
git remote add gajomi https://github.com/gajomi/julia
git fetch --all
git checkout structuredmatricprinting
git rebase -i JuliaLang/master
# squash into a smaller number of commits,
# resolve merge conflict, git add, git rebase --continue
# check that git log looks right
git push gajomi +structuredmatrixprinting
```

gajomi committed:
added methods to allow for overriding of element printing in arrays

gajomi commented:
Thanks. I am getting conflicts now. Will get it fixed up and push soon.

tkelman commented on test/show.jl:
```diff
@@ -338,3 +338,13 @@ end
 
 # Issue 14121
 @test_repr "(A'x)'"
+=======
```
merge artifact

gajomi commented:
Is there anything else that needs to be done to merge? I am interested in cooking up a separate pull request to document this behavior in the relevant section(s) of the manual, but wanted to make sure that the feature itself would merged first.

tkelman commented on base/show.jl:
```diff
@@ -1050,6 +1050,17 @@ function alignment(
 """
+Unexported convenience function used in body of `replace_in_print_matrix`
+methods. By default returns a string of the same width as original with a
+centered cdot, used in printing of structure zeros of structured matrices.
```
"structural zeros" ?

tkelman commented on base/show.jl:
```diff
@@ -1050,6 +1050,17 @@ function alignment(
+Unexported convenience function used in body of `replace_in_print_matrix`
+methods. By default returns a string of the same width as original with a
+centered cdot, used in printing of structure zeros of structured matrices.
+Accept keywrod args `c` for alternate single character marker.
```
keyword

tkelman commented on base/show.jl:
```diff
@@ -1072,7 +1082,8 @@ function print_matrix_row(io::IO,
         r = repeat(" ", A[k][2]-a[2])
-        print(io, l, sx, r)
+        prettysx = replace_in_print_matrix(X,i,j,sx)
+        print(io, l,prettysx,r)
```
keep the spaces after commas

tkelman commented:
Aside from typos / minor formatting nits, I'm in favor. Thanks for the bump.

ViralBShah commented:
Should we backport?

tkelman commented:
Don't think so. Fine for master but display stylistic changes don't seem like the best things to backport.

gajomi commented:
Well I ended up getting caught by winter holiday stuff before I could fix the various typos. I just did and now ci says its unmergeable. I couldn't see a link to any further information, but perhaps I have to rebase again?

gajomi commented:
Well, I ended up getting sucked in to holiday stuff before I could get to this, but the typo/style issue are fixed now. Should be ready to merge.
